plugins/inputs/sflow/types/decoder.go

```diff
@@ -0,0 +1,285 @@
+	Flags                uint8
+	FragmentOffset       uint16
+	TTL                  uint8
+	Protocol             uint8 // https://en.wikipedia.org/wiki/List_of_IP_protocol_numbers
+	HeaderChecksum       uint16
+	SourceIP             [4]byte
+	DestIP               [4]byte
+	ProtocolHeader       ProtocolHeader
+}
+
+func (h IPV4Header) GetTags() map[string]string {
+	var t map[string]string
+	if h.ProtocolHeader != nil {
+		t = h.ProtocolHeader.GetTags()
+	} else {
+		t = map[string]string{}
+	}
+	t["ip_dscp"] = strconv.FormatUint(uint64(h.DSCP), 10)
+	t["ip_ecn"] = strconv.FormatUint(uint64(h.ECN), 10)
+	t["src_ip"] = net.IP(h.SourceIP[:]).String()
+	t["dst_ip"] = net.IP(h.DestIP[:]).String()
+	return t
+}
+func (h IPV4Header) GetFields() map[string]interface{} {
+	var f map[string]interface{}
+	if h.ProtocolHeader != nil {
+		f = h.ProtocolHeader.GetFields()
+	} else {
+		f = map[string]interface{}{}
+	}
+	f["ip_flags"] = h.Flags
+	f["ip_fragment_offset"] = h.FragmentOffset
+	f["ip_total_length"] = h.TotalLength
+	f["ip_ttl"] = h.TTL
+	return f
+}
+
+// https://en.wikipedia.org/wiki/IPv6_packet
+type IPV6Header struct {
+	DSCP            uint8 `tag:"ip_dscp"`
+	ECN             uint8 `tag:"ip_ecn"`
+	PayloadLength   uint16
+	NextHeaderProto uint8 // tcp/udp?
+	HopLimit        uint8
+	SourceIP        [16]byte
+	DestIP          [16]byte
+	ProtocolHeader  ProtocolHeader
+}
+
+func (h IPV6Header) GetTags() map[string]string {
+	var t map[string]string
+	if h.ProtocolHeader != nil {
+		t = h.ProtocolHeader.GetTags()
+	} else {
+		t = map[string]string{}
+	}
+	t["ip_dscp"] = strconv.FormatUint(uint64(h.DSCP), 10)
+	t["ip_ecn"] = strconv.FormatUint(uint64(h.ECN), 10)
+	t["src_ip"] = net.IP(h.SourceIP[:]).String()
+	t["dst_ip"] = net.IP(h.DestIP[:]).String()
+	return t
+}
+func (h IPV6Header) GetFields() map[string]interface{} {
+	var f map[string]interface{}
+	if h.ProtocolHeader != nil {
+		f = h.ProtocolHeader.GetFields()
+	} else {
+		f = map[string]interface{}{}
+	}
+	// f["ip_total_length"] = h.PayloadLength // is this the same?
+	return f
+}
+
+// https://en.wikipedia.org/wiki/Transmission_Control_Protocol
+type TCPHeader struct {
+	SourcePort       uint16
+	DestinationPort  uint16
+	Sequence         uint32
+	AckNumber        uint32
+	TCPHeaderLength  uint8
+	Flags            uint16
+	TCPWindowSize    uint16
+	Checksum         uint16
+	TCPUrgentPointer uint16
+}
+
+func (h TCPHeader) GetTags() map[string]string {
+	t := map[string]string{
+		"dst_port": strconv.FormatUint(uint64(h.DestinationPort), 10),
+		"src_port": strconv.FormatUint(uint64(h.SourcePort), 10),
+	}
+	return t
+}
+func (h TCPHeader) GetFields() map[string]interface{} {
+	return map[string]interface{}{
+		"tcp_header_length":  h.TCPHeaderLength,
+		"tcp_urgent_pointer": h.TCPUrgentPointer,
+		"tcp_window_size":    h.TCPWindowSize,
+	}
+}
+
+type UDPHeader struct {
+	SourcePort      uint16
+	DestinationPort uint16
+	UDPLength       uint16
+	Checksum        uint16
```

Comment:
I noticed Checksum isn't used, should we remove it?

Reply:
Not sure if it might be useful to someone.